src/client_side_reply.cc

```diff
@@ -316,13 +316,17 @@ clientReplyContext::processExpired()
                                  http->log_uri, http->request->flags, http->request->method);
         /* NOTE, don't call StoreEntry->lock(), storeCreateEntry() does it */
 
-        if (collapsingAllowed) {
-            debugs(88, 5, "allow other revalidation requests to collapse on " << *entry);
-            Store::Root().allowCollapsing(entry, http->request->flags,
-                                          http->request->method);
-            collapsedRevalidation = crInitiator;
-        } else {
-            collapsedRevalidation = crNone;
+        if (collapsingAllowed)
+            Store::Root().allowCollapsing(entry, http->request->flags, http->request->method);
+
+        collapsedRevalidation = crNone;
+        if (!EBIT_TEST(entry->flags, KEY_PRIVATE)) {
```

Comment:
Look at storeCreateEntry(). The entry can become 'public' independently from collapsing feature.

Reply:
If storeCreateEntry() creates a public entry for a writer, then storeCreateEntry() should be responsible for creating a Transients entry to maintain the invariant.

Reply:
> If storeCreateEntry() creates a public entry for a writer

It is hardly possible to follow your suggestion.There are lots of storeCreateEntry() uses, some of them unrelated to our HTTP transactions dealing with Transients. As I said in another thread, it may not be enough information to decide that we have writer at storeCreateEntry() call time, or that the entry is not yet writer(for collapsing mode) or will fail to become writer in some cases.

Reply:
I did not mean to suggest something that was impossible or required widespread changes here. Let me rephrase in hope to minimize misunderstanding:

The following sequence of doing a heavy step (action1) and then immediately undoing it (action2) feels wrong. Please find a way to avoid this do-undo sequence.

```
e = storeCreateEntry(); // may call e->makePublic(); -- heavy action1
if (collapsingAllowed)
    allowCollapsing(e);
if (e->isPublic() && !e->preparePublic())
    e->makePrivate(); // action2 -- undoing heavy action1
```